src/Features/CSharp/Portable/CodeRefactorings/SyncNamespace/CSharpSyncNamespaceService.cs

```diff
@@ -0,0 +1,348 @@
+
+                if (!containsPartial)
+                {
+                    return triggeringNode;
+                }
+            }
+
+            return default;
+
+            SyntaxNode GetTriggeringNode(CompilationUnitSyntax compUnit, int pos)
+            {
+                var namespaceDecls = compilationUnit.DescendantNodes().OfType<NamespaceDeclarationSyntax>().ToImmutableArray();
+
+                if (namespaceDecls.Length == 1 && compilationUnit.Members.Count == 1)
+                {
+                    var namespaceDeclaration = namespaceDecls.Single();
+                    Debug.Assert(namespaceDeclaration == compilationUnit.Members.Single());
+
+                    if (namespaceDeclaration.Name.Span.IntersectsWith(position)
+                        && namespaceDeclaration.Name.GetDiagnostics().All(diag => diag.DefaultSeverity != DiagnosticSeverity.Error))
+                    {
+                        return namespaceDeclaration;
+                    }
+                }
+                else if (namespaceDecls.Length == 0)
+                {
+                    var firstMemberDeclarationName = compilationUnit.Members.FirstOrDefault().GetNameToken();
+
+                    if (firstMemberDeclarationName != default 
+                        && firstMemberDeclarationName.Span.IntersectsWith(position))
+                    {
+                        return compilationUnit;
+                    }
+                }
+
+                return null;
+            }
+        }
+
+        /// <summary>
+        /// Try to change the namespace declaration based on the refacoring rules:
+        ///     - if neither declared and target namespace are "" (i.e. global namespace),
+        ///     then we try to change the name of the namespace.
+        ///     - if declared namespace is "", then we try to move all types declared 
+        ///     in global namespace in the document into a new namespace declaration.
+        ///     - if target namespace is "", then we try to move all members in declared 
+        ///     namespace to global namespace (i.e. remove the namespace declaration).    
+        /// </summary>
+        protected override CompilationUnitSyntax ChangeNamespaceDeclaration(
+            CompilationUnitSyntax compilationUnit, 
+            ImmutableArray<string> declaredNamespaceParts, 
+            ImmutableArray<string> targetNamespaceParts)
+        {
+            Debug.Assert(!declaredNamespaceParts.IsDefault && !targetNamespaceParts.IsDefault);
+
+            // Move everything from global namespace to a namespace declaration
+            if (declaredNamespaceParts.Length == 1 && declaredNamespaceParts[0].Length == 0)
+            {
+                var targetNamespaceDecl = SyntaxFactory.NamespaceDeclaration(
+                    name: CreateNameSyntax(targetNamespaceParts, aliasQualifier: null, targetNamespaceParts.Length - 1)
+                            .WithAdditionalAnnotations(WarningAnnotation),
```

Comment:
in test, I didn't see the warning. might want to add the warning in test as well? so people can see when warning might be shown to users?

Reply:
Good point, will do